rfcs/20221124-monorepo.md

```diff
@@ -0,0 +1,179 @@
+
+### Step 0: Prerequisites
+
+#### Get the timing right
+
+The first step will be to release the frontend, call a code freeze and pause work on it. This is to prevent the frontend repo from continuing to drift as we merge a snapshot of it with the API.
+
+This can prove difficult given how productive our team is, so we will need to channel this productivity towards the catalog in the meantime. I can foresee the end-to-end migration taking one week (ideal scenario) to one fortnight (worst case scenario).
+
+### Step 1: Merge with histories
+
+This is a quick process.
+
+1. Move the entire content of frontend inside a `frontend/` directory, except the following top-level files and folders. Please comment if you can add to this list.
+
+   - `.github/`
+   - `.editorconfig`
+   - `justfile`
+   - `.pre-commit-config.yaml`
+   - `prettier.config.js`
+   - `.prettierignore`
+   - `.eslintrc.js` (need to update references in nested `eslintrc.js` files)
+   - `.eslintignore`
+   - <s>`.gitignore`</s> (better to move it into the `frontend/` directory and update some absolute paths)
+
+1. Create the final commit on the `WordPress/openverse-frontend` repo. After the merge we might want to add a notice about the migration to the `README.md` file but GitHub's built-in archival process could suffice here.
+
+1. Merge this repo's `main` branch into the `WordPress/openverse-api` repo's `main` branch (see Git docs for `--allow-unrelated-histories`). There will be some conflicts but they will be small and infrequent. [[implementation details](#conflict-resolution)]
+
+1. Create "stack: \*" labels to help with issue and PR management. Spoiler/foreshadowing: these labels will be used for more things later.
+
+1. Migrate issues from `WordPress/openverse-frontend` to `WordPress/openverse-api`. @obulat's has done prior work in this department (when we migrated from CC Search to Openverse) but that might not be as useful because in this case, we can directly transfer the issues, retaining all their comments. Apply the "stack: frontend" label to them. [[implementation details](#issue-transfer)]
+
+With this done, we can archive the frontend repo.
+
+#### Conflict resolution
+
+The following conflicts may occur during merge.
+
+- `.prettierignore`: concatenate
+- `.pre-commit-config.yaml`: use from [dhruvkb/monopenverse](https://github.com/dhruvkb/monopenverse)
+- Workflows can conflict but they can be renamed and kept alongside each other, _for now_.
+
+#### Issue transfer
+
+As far as I can tell, issue transfer can only be performed via the GitHub GraphQL API ([docs](https://docs.github.com/en/graphql/reference/mutations#transferissue)) and not via the REST API. From my limited testing, transferred issues seem to retain labels (provided they exist in the target repo).
```

Comment:
I wrote a script in my previous monorepo exploration to transfer issues from one repo (`openverse/service` is the sample name I use, `WordPress/openverse-frontend` would be a concrete example, into another repo:

`gh api -X GET repos/openverse/service/issues -f pulls=false -f state=all --jq '.[].number' --paginate | xargs -n 1 -I num gh issue transfer num openverse/monorepo -R openverse/service`

In testing it worked quite well, and quickly.

Reply:
Thanks! I've included the command in the RFC itself in f2a9d81.